File: projects/five-cycle-double-cover/scratch/verify_five_point_triangle_list_lift.py
#!/usr/bin/env python3
"""Literal audit of the five-point restriction of Fano triangle labels.

The eight points are the integers 0,...,7 with bitwise xor.  A Fano line
is represented by its four-element vector subspace, including zero.  A
local coordinate triangle for that line is a three-element subset of one
of its two affine cosets.
"""

from __future__ import annotations

from collections import Counter
from itertools import combinations
import json


POINTS = frozenset(range(8))


def xor_sum(values: frozenset[int] | set[int]) -> int:
    result = 0
    for value in values:
        result ^= value
    return result


def fano_planes() -> tuple[frozenset[int], ...]:
    planes = {
        frozenset((0, first, second, first ^ second))
        for first in range(1, 8)
        for second in range(first + 1, 8)
    }
    assert len(planes) == 7
    return tuple(sorted(planes, key=lambda plane: tuple(sorted(plane))))


def affine_triangles(plane: frozenset[int]) -> tuple[frozenset[int], ...]:
    translate = next(point for point in POINTS if point not in plane)
    cosets = (plane, frozenset(point ^ translate for point in plane))
    triangles = {
        frozenset(coset - {omitted})
        for coset in cosets
        for omitted in coset
    }
    assert len(triangles) == 8
    return tuple(sorted(triangles, key=lambda triangle: tuple(sorted(triangle))))


def pair_with_difference(
    triangle: frozenset[int], difference: int
) -> frozenset[int]:
    pairs = [
        frozenset((left, right))
        for left, right in combinations(sorted(triangle), 2)
        if left ^ right == difference
    ]
    assert len(pairs) == 1
    return pairs[0]


def main() -> int:
    planes = fano_planes()
    five_sets = tuple(frozenset(values) for values in combinations(range(8), 5))
    assert len(five_sets) == 56

    line_case_counts = Counter()
    total_allowed_triangles = 0
    for five_set in five_sets:
        forbidden = POINTS - five_set
        singleton = xor_sum(set(forbidden))
        assert singleton in five_set

        forbidden_plane = frozenset(forbidden | {singleton})
        special = frozenset(point ^ singleton for point in forbidden_plane)
        assert special in planes
        other_coset = five_set - {singleton}
        assert len(other_coset) == 4
        assert frozenset(point ^ singleton for point in other_coset) != special

        special_cases = 0
        for plane in planes:
            allowed = [
                triangle
                for triangle in affine_triangles(plane)
                if triangle <= five_set
            ]
            expected = 4 if plane == special else 1
            assert len(allowed) == expected
            line_case_counts[(plane == special, len(allowed))] += 1
            total_allowed_triangles += len(allowed)

            if plane == special:
                special_cases += 1
                assert all(triangle <= other_coset for triangle in allowed)
                for triangle in allowed:
                    for difference in plane - {0}:
                        assert pair_with_difference(
                            triangle, difference
                        ) <= other_coset
                continue

            triangle = allowed[0]
            assert singleton in triangle
            intersection = (plane & special) - {0}
            assert len(intersection) == 1
            common_value = next(iter(intersection))
            assert pair_with_difference(triangle, common_value) <= other_coset
            for difference in plane - {0, common_value}:
                pair = pair_with_difference(triangle, difference)
                assert singleton in pair
                assert len(pair & other_coset) == 1
        assert special_cases == 1

    result = {
        "schema": "five-cdc-five-point-triangle-list-v1",
        "five_point_sets": len(five_sets),
        "fano_lines": len(planes),
        "line_cases": len(five_sets) * len(planes),
        "special_line_cases": line_case_counts[(True, 4)],
        "nonspecial_line_cases": line_case_counts[(False, 1)],
        "allowed_local_triangles": total_allowed_triangles,
        "result": (
            "For every five-point subset of F2^3, exactly one Fano line "
            "has four admitted local coordinate triangles and each of the "
            "other six lines has exactly one."
        ),
    }
    print(json.dumps(result, sort_keys=True, separators=(",", ":")))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: projects/five-cycle-double-cover/scratch/verify_five_point_triangle_list_lift.mjs
#!/usr/bin/env node
// Independent bit-mask replay of the five-point/Fano-line list theorem.

function bits(mask) {
  const result = [];
  for (let point = 0; point < 8; point += 1) {
    if ((mask >>> point) & 1) result.push(point);
  }
  return result;
}

function xor(values) {
  return values.reduce((total, value) => total ^ value, 0);
}

function chooseMasks(size) {
  const result = [];
  for (let mask = 0; mask < 256; mask += 1) {
    if (bits(mask).length === size) result.push(mask);
  }
  return result;
}

const triples = chooseMasks(3);
const fiveSets = chooseMasks(5);
const planeByTriple = new Map();
const planes = new Set();

for (const triple of triples) {
  const [a, b, c] = bits(triple);
  const directions = [0, a ^ b, a ^ c, b ^ c];
  const plane = directions.reduce((mask, point) => mask | (1 << point), 0);
  planeByTriple.set(triple, plane);
  planes.add(plane);
}

if (planes.size !== 7 || triples.length !== 56 || fiveSets.length !== 56) {
  throw new Error("wrong ambient enumeration");
}

let specialCases = 0;
let ordinaryCases = 0;
let admitted = 0;

for (const five of fiveSets) {
  const forbidden = 0xff ^ five;
  const singleton = xor(bits(forbidden));
  if (!((five >>> singleton) & 1)) {
    throw new Error("affine-plane fourth point is not admitted");
  }
  const special = planeByTriple.get(forbidden);
  const otherCoset = five ^ (1 << singleton);

  for (const plane of planes) {
    const allowed = triples.filter(
      (triple) => planeByTriple.get(triple) === plane && (triple & ~five) === 0,
    );
    const expected = plane === special ? 4 : 1;
    if (allowed.length !== expected) throw new Error("wrong local list size");
    admitted += allowed.length;

    if (plane === special) {
      specialCases += 1;
      for (const triple of allowed) {
        if ((triple >>> singleton) & 1) {
          throw new Error("special triangle uses singleton");
        }
        if (triple & ~otherCoset) throw new Error("special triangle leaves coset");
      }
      continue;
    }

    ordinaryCases += 1;
    const triangle = allowed[0];
    if (!((triangle >>> singleton) & 1)) {
      throw new Error("ordinary triangle misses singleton");
    }
    const common = bits(plane & special).filter((point) => point !== 0);
    if (common.length !== 1) throw new Error("wrong line intersection");

    const trianglePoints = bits(triangle);
    for (const difference of bits(plane).filter((point) => point !== 0)) {
      const pairs = [];
      for (let i = 0; i < 3; i += 1) {
        for (let j = i + 1; j < 3; j += 1) {
          if ((trianglePoints[i] ^ trianglePoints[j]) === difference) {
            pairs.push([trianglePoints[i], trianglePoints[j]]);
          }
        }
      }
      if (pairs.length !== 1) throw new Error("difference has wrong multiplicity");
      const includesSingleton = pairs[0].includes(singleton);
      if (includesSingleton === (difference === common[0])) {
        throw new Error("pair-shape classification failed");
      }
    }
  }
}

const result = {
  schema: "five-cdc-five-point-triangle-list-independent-v1",
  five_point_sets: fiveSets.length,
  fano_lines: planes.size,
  special_line_cases: specialCases,
  nonspecial_line_cases: ordinaryCases,
  allowed_local_triangles: admitted,
  result: "PASS",
};
process.stdout.write(`${JSON.stringify(result)}\n`);
